tools/autodeployment/builds.js

```diff
@@ -0,0 +1,130 @@
+const { logger, createError } = require('@senecacdot/satellite');
+const shell = require('shelljs');
+const mergeStream = require('merge-stream');
+
+class Build {
+  constructor(type, githubData) {
+    this.type = type;
+    this.githubData = githubData;
+    this.startedDate = new Date();
+  }
+
+  finish(code) {
+    logger.debug({ code }, 'build finished');
+    this.stopDate = new Date();
+    this.code = code;
+
+    // Drop the output stream
+    if (this.out) {
+      this.out = null;
+    }
+    // Drop the deploy process
+    if (this.proc) {
+      this.proc = null;
+    }
+  }
+
+  toJSON() {
+    const build = {
+      type: this.type,
+      githubData: this.githubData,
+      startedDate: this.startedDate,
+    };
+
+    if (this.stopDate) {
+      build.stopDate = this.stopDate;
+    }
+    if (this.code) {
+      build.code = this.code;
+    }
+
+    return build;
+  }
+}
+
+// Simple queue of builds we have to do
+const builds = {
+  // Whatever built most recently
+  previous: null,
+  // Whatever we're building now
+  current: null,
+  // Whatever we're building next
+  pending: null,
+};
+
+function run() {
+  // Run the next build, unless we're already running one
+  if (builds.current) {
+    logger.debug('run() called, but build already running, skipping.');
+    return;
+  }
+
+  // See if there's anything waiting to be built
+  if (!builds.pending) {
+    logger.debug('run() called, but no pending build to run, skipping.');
+    return;
+  }
+
+  // Promote the next build to current and start it
+  const build = builds.pending;
+  builds.current = build;
+  builds.pending = null;
+
+  logger.debug('run() called, starting pending build');
+  build.proc = shell.exec(`./deploy.sh ${build.type}`, { silent: true }, (code) => {
+    build.finished(code);
```

Comment:
This happens after a request was processed:
```sh
22:56:15 ✨ Server listening on port 4000
22:59:56 ✨ request completed POST 200 / 29ms
23:03:00 ✨ request completed POST 200 / 2ms
23:03:26 ✨ request completed GET 200 /log 6.4s
/usr/share/telescope_project/autodeployment/builds.js:75
    build.finished(code);
          ^

TypeError: build.finished is not a function
    at /usr/share/telescope_project/autodeployment/builds.js:75:11
    at /usr/share/telescope_project/autodeployment/node_modules/shelljs/src/exec.js:127:9
```

Reply:
Fixed